eos-updater-flatpak-installer/main.c

```diff
@@ -0,0 +1,383 @@
+  g_autofree gchar *counter_as_string = g_strdup_printf ("%lu", action->order);
+
+  if (!g_file_make_directory_with_parents (parent, NULL, &local_error))
+    {
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return FALSE;
+        }
+
+      g_clear_error (&local_error);
+    }
+
+  return g_file_replace_contents (counter_file,
+                                  counter_as_string,
+                                  strlen (counter_as_string),
+                                  NULL,
+                                  FALSE,
+                                  G_FILE_CREATE_NONE,
+                                  NULL,
+                                  NULL,
+                                  error);
+}
+
+static void
+update_counter_complain_on_error (GFile                   *counter_file,
+                                  FlatpakRemoteRefAction  *action,
+                                  const gchar             *name)
+{
+  g_autoptr(GError) error = NULL;
+
+  if (!update_counter (counter_file, action, &error))
+    complain_about_failure_to_update_system_installation_counter (name,
+                                                                  error);
+}
+
+static gboolean
+apply_flatpak_ref_actions (GHashTable *table, GError **error)
+{
+  g_autoptr(FlatpakInstallation) installation = eos_updater_get_flatpak_installation (NULL, error);
+  g_autoptr(GError) local_error = NULL;
+  gpointer key;
+  gpointer value;
+  GHashTableIter hash_iter;
+
+  if (!installation)
+    return FALSE;
+
+  g_hash_table_iter_init (&hash_iter, table);
+
+  while (g_hash_table_iter_next (&hash_iter, &key, &value))
+    {
+      const gchar *name = key;
+      GPtrArray *pending_actions = value;
+      g_autofree gchar *counter_path = g_build_filename (eos_updater_util_pending_flatpak_deployments_state_path (),
+                                                         name,
+                                                         NULL);
+      g_autoptr(GFile) counter_file = g_file_new_for_path (counter_path);
+      gsize i;
+      FlatpakRemoteRefAction *last_successful_action = NULL;
+
+      for (i = 0 ; i < pending_actions->len; ++i)
+        {
+          FlatpakRemoteRefAction *pending_action = g_ptr_array_index (pending_actions, i);
+
+          if (!perform_action (installation, pending_action, error))
+            {
+              /* If we fail, we should still update the state of the counter
+               * to the last successful before we get out, this is to ensure
+               * that we don't perform the same action again next time */
+              update_counter_complain_on_error (counter_file,
+                                                last_successful_action,
+                                                name);
+              last_successful_action = NULL;
+              break;
```

Comment:
A consequence of breaking here and the following logic is that in this code path the function will `return TRUE` with the error set, and an error did actually happen. I believe you intend to `return FALSE` when this is hit.

Reply:
Ugh, yeah, that is true. The good news is that we can just `return FALSE`.